File: go/internal/agent/setup/session.go
package setup

import (
	"context"
	"fmt"

	"github.com/glebarez/sqlite"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"

	"google.golang.org/adk/session"
	"google.golang.org/adk/session/database"

	"github.com/jkjamies/automation-agent/internal/config"
)

// NewSessionService builds the ADK session service for the configured backend. The
// session service holds the durable suspend/resume history that lets a parked fix run
// continue after a process restart. Keeping the sqlite/gorm imports here respects the
// ARCH boundary (infrastructure SDKs live under internal/agent/setup).
//
//	memory    -> in-process; tests and ephemeral local runs (today's behavior, default)
//	sqlite    -> file-backed via adk session/database; durable local runs
//	firestore -> cloud, via the custom Firestore-backed service; durable cloud runs
//
// Only the long-running fix loop needs durability; ephemeral one-shot runners
// (explore/analyze/triage) keep using an in-memory session via NewRunner.
func NewSessionService(ctx context.Context, cfg config.Config) (session.Service, error) {
	switch cfg.SessionBackend {
	case config.SessionMemory:
		return session.InMemoryService(), nil
	case config.SessionSQLite:
		// Silent GORM logger: the get-or-create path issues benign "record not found"
		// lookups for unset app/user state. Real failures still propagate as the
		// returned errors from the service methods, independent of this logger.
		svc, err := database.NewSessionService(sqlite.Open(cfg.SQLiteDSN), &gorm.Config{
			PrepareStmt: true,
			Logger:      logger.Default.LogMode(logger.Silent),
		})
		if err != nil {
			return nil, fmt.Errorf("sqlite session service: %w", err)
		}
		if err := database.AutoMigrate(svc); err != nil {
			return nil, fmt.Errorf("sqlite session migrate: %w", err)
		}
		return svc, nil
	case config.SessionFirestore:
		return NewFirestoreSessionService(ctx, cfg.FirestoreProject, cfg.FirestoreCollection)
	default:
		return nil, fmt.Errorf("unknown session backend %q", cfg.SessionBackend)
	}
}

File: javascript/src/agent/setup/session.ts
/**
 * The session-service factory: selects where the suspend/resume session history lives.
 *
 * The fix loop suspends on a long-running tool and resumes when CI reports; the session
 * service is what holds that conversation across the wait. `memory` (the default) keeps
 * it in-process — a restart drops parked runs; the durable backends persist it so a
 * parked run survives a restart. Kept in `setup` because it touches the ADK session
 * SDK (confined here by the arch tests).
 */
import { resolve } from 'node:path';

import { type BaseSessionService, DatabaseSessionService, InMemorySessionService } from '@google/adk';

import { type Config, SessionBackend } from '../../config/config';
import { FirestoreSessionService } from './session_firestore';

/** Build the ADK session service for the configured backend. */
export function newSessionService(cfg: Config): BaseSessionService {
  switch (cfg.sessionBackend) {
    case SessionBackend.Memory:
      return new InMemorySessionService();
    case SessionBackend.Sqlite:
      // ADK's database-backed session service over MikroORM; the `sqlite://` URI takes an
      // absolute path so it shares the configured file with the sqlite park store.
      return new DatabaseSessionService(`sqlite://${resolve(cfg.sqliteDsn)}`);
    case SessionBackend.Firestore:
      return new FirestoreSessionService(cfg.firestoreProject, cfg.firestoreCollection);
  }
}
